### PR TITLE
Fixes for UCR expressions in the DB

### PR DESCRIPTION
**Pull request by snopoke**
## Product Description
No changes to product features other than bug fixes.

## Technical Summary
There are two bug fixes in this PR:

**AttributeError**

`wrapped_definition` method expects a `FactoryContext` object, not an `EvaluationContext`: https://dimagi.sentry.io/share/issue/e3045f92b61849a2af70defda9fb0769/

**RecursionError**
`maximum recursion depth exceeded while calling a Python object`

This occurs when a `UCRExpression` object references another `UCRExpression` via a `named_expression`. When wrapping the first expression the FactoryContext attempts to load AND wrap all expressions which includes the one that is currently being wrapped. This causes a recursive loop.

To resolve this I have introduced `LazyExpressionWrapper` class which delays the wrapping of UCR expression specs until they are needed. This allows us to validate that an expression with the referenced name exists without having to wrap the referenced expression. There are two added benefits here: don't wrap expressions that aren't needed + don't raise errors from unrelated expressions.

This still left room for infinite recursion due to circular references. To detect this case I've refactored the FactoryContext to maintain a call stack and check for duplicates in the stack when resolving references.

Can be reviewed by commit.

## Feature Flag
`UCR_EXPRESSION_REGISTRY` (though the refactors will impact other configurable report code)

### Safety story
There is a good test suite and I've added other tests. The scope of these classes is relatively small so I'm confident in the changes to the public API of `FactoryContext`

### Automated test coverage
Existing + new tests

### QA Plan
None though I could be persuaded to put it on staging

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/apps/userreports/specs.py [hunk 1]
```diff
@@ -39,29 +46,62 @@ def filter_from_spec(self, spec):
...
+        extra_filters = {}
         if self.domain and toggles.UCR_EXPRESSION_REGISTRY.enabled(self.domain):
             from corehq.apps.userreports.models import UCRExpression
-            extra_expressions = UCRExpression.objects.get_wrapped_expressions_for_domain(self.domain, self)
-        return extra_expressions | self._named_expressions
+            extra_filters = UCRExpression.objects.get_wrapped_filters_for_domain(self.domain, self)
+        return extra_filters
+
+    def get_named_filter(self, name):
+        return self._get_named(name, self._named_filters | self._extra_filters, self._filter_stack)
```
Comment (esoergel) [inline corehq/apps/userreports/specs.py]: I notice this flips the order around the `|`, meaning `_extra_filters` will now override `_named_filters` where they conflict.  Was that change deliberate?

File: corehq/apps/userreports/specs.py [hunk 2]
```diff
@@ -39,29 +46,62 @@ def filter_from_spec(self, spec):
...
+    @memoized
+    def _extra_expressions(self):
+        extra_expressions = {}
+        if self.domain and toggles.UCR_EXPRESSION_REGISTRY.enabled(self.domain):
+            from corehq.apps.userreports.models import UCRExpression
+            extra_expressions = UCRExpression.objects.get_wrapped_expressions_for_domain(self.domain, self)
+        return extra_expressions
+
+    def get_named_expression(self, name):
+        return self._get_named(name, self._named_expressions | self._extra_expressions, self._expression_stack)
```
Comment (esoergel) [inline corehq/apps/userreports/specs.py]: This one likewise flips the ordering

**Review by esoergel — APPROVED** [review]
There's a possible behavior change I wasn't sure of, but otherwise this looks g2g

File: corehq/apps/userreports/specs.py
Comment (millerdev) [inline corehq/apps/userreports/specs.py:1]: "Can be reviewed by commit." in the description was kinda cruel, given that 5f520b956564869835022487e06e76c86ceaee4f invalidated most code comments on prior commits. I feel like I'm starting over with my review now.

**Review by millerdev — COMMENTED** [review]
I didn't finish reviewing this because I spent too much time on it and then got overwhelmed with the refactor commit. I'll try to come back to it tomorrow.

File: corehq/apps/userreports/models.py [hunk 3]
```diff
@@ -1486,10 +1486,13 @@ def __init__(self, expression, factory_context):
         self.wrapped_expression = None
 
     def __call__(self, *args, **kwargs):
+        self.unwrap()
+        return self.wrapped_expression(*args, **kwargs)
+
+    def unwrap(self):
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/apps/userreports/models.py:1482]: > Or possibly make it a `django.utils.functional.cached_property` named `wrapped_expression`

I like this suggestion too - it's used in one place to mutate the class and in the other place for the return value.  Making this a cached property would allow it to be used the same way in both contexts.

File: corehq/apps/userreports/specs.py [hunk 4]
```diff
@@ -39,29 +46,62 @@ def filter_from_spec(self, spec):
     @property
     @memoized
     def named_filters(self):
-        extra_filters = {}
-        if self.domain and toggles.UCR_EXPRESSION_REGISTRY.enabled(self.domain):
-            from corehq.apps.userreports.models import UCRExpression
-            extra_filters = UCRExpression.objects.get_wrapped_filters_for_domain(self.domain, self)
-        return extra_filters | self._named_filters
+        raise Exception("Use get_named_filter instead of named_filters directly.")
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/userreports/specs.py]: I added this as a stop gap during the refactor to make sure it wasn't still being called anywhere.

File: corehq/apps/userreports/specs.py [hunk 5]
```diff
@@ -39,29 +46,62 @@ def filter_from_spec(self, spec):
...
+        extra_expressions = {}
+        if self.domain and toggles.UCR_EXPRESSION_REGISTRY.enabled(self.domain):
+            from corehq.apps.userreports.models import UCRExpression
+            extra_expressions = UCRExpression.objects.get_wrapped_expressions_for_domain(self.domain, self)
+        return extra_expressions
+
+    def get_named_expression(self, name):
+        return self._get_named(name, self._named_expressions | self._extra_expressions, self._expression_stack)
+
+    def _get_named(self, name, named_dict, stack):
```
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/userreports/specs.py]: This was refactored later

File: corehq/apps/userreports/specs.py (hunk 2 above)
Comment (snopoke) [reply corehq/apps/userreports/specs.py]: 6b19be13f442c7ed44312517c28786a39816a9f9

File: corehq/apps/userreports/specs.py (hunk 1 above)
Comment (snopoke) [reply corehq/apps/userreports/specs.py]: 6b19be13f442c7ed44312517c28786a39816a9f9

**Review by esoergel — APPROVED** [review]

File: corehq/apps/userreports/specs.py
Comment (snopoke) [reply corehq/apps/userreports/specs.py:1]: Apologies for that. I said "by commit" because the logic changes can be hard to see in amongst the refactor.

File: corehq/apps/userreports/specs.py [hunk 6]
```diff
@@ -20,20 +19,70 @@ def TypeProperty(value):
...
+    def __init__(self, factory_context, expressions, domain=None):
+        self.factory_context = weakref.proxy(factory_context)
+        self.expressions = expressions
+        self.domain = domain
+        self.stack = []
+
+    def replace(self, expressions):
+        self.expressions = expressions
+
+    def get(self, name):
```
Comment (millerdev) [inline corehq/apps/userreports/specs.py:35]: This could be renamed to `__getitem__` to preserve the original dict item lookup inteface of `FactoryContext.named_expressions` and `FactoryContext.named_filters`.

File: corehq/apps/userreports/models.py [hunk 7]
```diff
@@ -466,24 +466,14 @@ def _get_domain_filter_spec(self):
...
-                except BadSpecError as bad_spec_error:
-                    # maybe a nested name resolution issue, try again on the next pass
-                    spec_error = bad_spec_error
-            if number_generated == 0 and named_expression_specs:
-                # we unsuccessfully generated anything on this pass and there are still unresolved
-                # references. we have to fail.
-                assert spec_error is not None
-                raise spec_error
+        factory_context = FactoryContext.empty(self.domain)
+        for name, expression in list(named_expression_specs.items()):
```
Comment (millerdev) [inline corehq/apps/userreports/models.py:470]: Why is `list()` needed here?

File: corehq/apps/userreports/filters/factory.py [hunk 8]
```diff
@@ -71,7 +71,7 @@ def _build_boolean_expression_filter(spec, factory_context):
 def _build_named_filter(spec, factory_context):
     wrapped = NamedFilterSpec.wrap(spec)
     try:
-        filter = factory_context.named_filters[wrapped.name]
+        filter = factory_context.get_named_filter(wrapped.name)
     except KeyError as e:
```
Comment (millerdev) [inline corehq/apps/userreports/filters/factory.py:75]: Does `get_named_filter` raise `KeyError` (as the dict lookup did) when the lookup fails? I think it does not.

File: corehq/apps/userreports/specs.py [hunk 9]
```diff
@@ -18,15 +19,70 @@ def TypeProperty(value):
...
+        with self._search_scope(name):
+            from corehq.apps.userreports.models import LazyExpressionWrapper
+            all_expressions = self._get_db_expressions() | self.expressions
+            expr = all_expressions.get(name)
+            if not expr:
+                raise BadSpecError(gettext("Couldn't find named expression with name: {name}").format(name=name))
```
Comment (millerdev) [inline corehq/apps/userreports/specs.py:41]: This error message always uses the word "expression", but sometimes it's a _filter_ that's being looked up, right?

Another nudge that this `get` method should be renamed to `__getitem__` and it should raise `KeyError` here.

File: corehq/apps/userreports/specs.py (hunk 6 above)
Comment (snopoke) [reply corehq/apps/userreports/specs.py:35]: This is only called from within the `FactoryContext`. Users of the context call `get_named_filter` or `get_named_expression`. I specifically didn't make this mimic dict item lookup to make it easier to trace calls in code and to avoid assumptions that the underlying object is a dict.

File: corehq/apps/userreports/models.py (hunk 7 above)
Comment (snopoke) [reply corehq/apps/userreports/models.py]: It isn't, just a holdover from the previous code. c00805aec479151cb86bab217a9e139153eb9f75

File: corehq/apps/userreports/filters/factory.py (hunk 8 above)
Comment (snopoke) [reply corehq/apps/userreports/filters/factory.py]: Nope: 51c8572a3d643a292262197a05716c17476d8726

File: corehq/apps/userreports/specs.py (hunk 9 above)
Comment (snopoke) [reply corehq/apps/userreports/specs.py]: I debated about correcting that but it didn't seem necessary as the names provide clear indication of what was missing. Raising a `KeyError` would require handling the error by the callers which just convert it to another error anyway. But, I can push the error handling up one level to the `FactoryContext` which solves both issues: de4b0b5f35f29ee579b66a3179fef2d87b909d47

File: corehq/apps/userreports/models.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (snopoke) [reply corehq/apps/userreports/models.py]: 1160e29e9f6b520f33fca98cabcc162fd7303498

File: corehq/apps/userreports/models.py [hunk 10]
```diff
@@ -6,13 +6,13 @@
 from collections import namedtuple
 from copy import copy, deepcopy
 from datetime import datetime, timedelta
+from functools import cached_property
```
Comment (millerdev) [inline corehq/apps/userreports/models.py:9]: I didn't realize this had made it into the standard library. Nice!

**Review by millerdev — APPROVED** [review]

File: corehq/apps/userreports/models.py (hunk 10 above)
Comment (esoergel) [reply corehq/apps/userreports/models.py:9]: oh cool, me neither!  Definitely feels like a good fit for it
